src/workerd/api/urlpattern-standard.c++

```diff
@@ -0,0 +1,253 @@
+
+  KJ_IF_SOME(mi, maybeInput) {
+    KJ_SWITCH_ONEOF(mi) {
+      KJ_CASE_ONEOF(str, kj::String) {
+        input = std::string_view(str.begin(), str.size());
+      }
+      KJ_CASE_ONEOF(init, URLPatternInit) {
+        input = init.toAdaType();
+      }
+    }
+  } else {
+    input = ada::url_pattern_init{};
+  }
+
+  KJ_IF_SOME(b, maybeBase) {
+    KJ_SWITCH_ONEOF(b) {
+      KJ_CASE_ONEOF(str, kj::String) {
+        base = std::string_view(str.begin(), str.size());
+      }
+      KJ_CASE_ONEOF(o, URLPatternOptions) {
+        options = o.toAdaType();
+      }
+    }
+  }
+
+  if (!options.has_value()) {
+    KJ_IF_SOME(o, maybeOptions) {
+      options = o.toAdaType();
+    }
+  }
+
+  std::string_view* base_opt = base ? &(*base) : nullptr;
+  auto result =
+      ada::parse_url_pattern<URLPatternRegexEngine>(input, base_opt, options ? &*options : nullptr);
+  if (result.has_value()) {
+    return jsg::alloc<URLPattern>(std::move(*result));
+  }
+
+  JSG_FAIL_REQUIRE(TypeError, "Failed to construct URLPattern");
+}
+
+bool URLPattern::test(jsg::Optional<kj::OneOf<kj::String, URLPatternInit>> maybeInput,
+    jsg::Optional<kj::String> maybeBase) {
+  ada::result<bool> result;
+  std::optional<std::string_view> base{};
+
+  KJ_IF_SOME(b, maybeBase) {
+    base = std::string_view(b.begin(), b.size());
+  }
+
+  std::string_view* base_ptr = base ? &*base : nullptr;
+
+  KJ_IF_SOME(mi, maybeInput) {
+    KJ_SWITCH_ONEOF(mi) {
+      KJ_CASE_ONEOF(str, kj::String) {
+        result = inner.test(std::string_view(str.begin(), str.size()), base_ptr);
+      }
+      KJ_CASE_ONEOF(pi, URLPattern::URLPatternInit) {
+        result = inner.test(pi.toAdaType(), base_ptr);
+      }
+    }
+  } else {
+    result = inner.test(ada::url_pattern_init{}, base_ptr);
+  }
+
+  JSG_REQUIRE(result.has_value(), TypeError, "Failed to test URLPattern");
+
+  return *result;
+}
+
+kj::Maybe<URLPattern::URLPatternResult> URLPattern::exec(jsg::Lock& js,
+    jsg::Optional<kj::OneOf<kj::String, URLPatternInit>> maybeInput,
+    jsg::Optional<kj::String> maybeBase) {
+  ada::result<std::optional<ada::url_pattern_result>> result;
+  std::optional<std::string_view> base_url{};
+  std::string base_url_base;
+
+  KJ_IF_SOME(b, maybeBase) {
+    base_url_base = std::string(b.begin(), b.size());
+    base_url = std::string_view(base_url_base);
+  }
+
+  KJ_IF_SOME(mi, maybeInput) {
+    KJ_SWITCH_ONEOF(mi) {
+      KJ_CASE_ONEOF(str, kj::String) {
+        result = inner.exec(
+            std::string_view(str.begin(), str.size()), base_url ? &(*base_url) : nullptr);
+      }
+      KJ_CASE_ONEOF(pi, URLPattern::URLPatternInit) {
+        result = inner.exec(pi.toAdaType(), base_url ? &(*base_url) : nullptr);
+      }
+    }
+  } else {
+    result = inner.exec(ada::url_pattern_init{}, base_url ? &(*base_url) : nullptr);
+  }
+
+  // If result does not exist, we should throw.
+  JSG_REQUIRE(result.has_value(), TypeError, "Failed to exec URLPattern");
```

Comment:
Does ada provide us with any more information regarding the failure type?

Reply:
URLPattern spec only returns TypeError and doesnt' mention any other error type. We can change the implementation and have unique error messages but it's not standardized, unfortunately.

Reply:
Resolving for now. We can go back to this in the future. cc @jasnell